include/dali/core/mm/composite_resource.h

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_MM_COMPOSITE_RESOURCE_H_
+#define DALI_CORE_MM_COMPOSITE_RESOURCE_H_
+
+#include <type_traits>
+#include <memory>
+#include <tuple>
+#include <utility>
+#include "dali/core/mm/memory_resource.h"
+
+namespace dali {
+namespace mm {
+
+namespace detail {
+
+template <memory_kind kind, typename Context>
+memory_resource<kind, Context> &GetResourceInterface(const memory_resource<kind, Context> &);
+
+template <memory_kind kind>
+async_memory_resource<kind> &GetResourceInterface(const async_memory_resource<kind> &);
+
+template <typename Resource>
+using resource_interface_t = std::remove_reference_t<
+    decltype(GetResourceInterface(std::declval<Resource>()))>;
+
+
+template <typename Interface, typename Resource, typename...Extra>
+class CompositeResourceBase : public Interface {
+ public:
+  CompositeResourceBase() = default;
+
+  template <typename... ExtraArgs>
+  CompositeResourceBase(std::shared_ptr<Resource> resource,
+                    ExtraArgs... extra)
```

Comment:
Why it can't be:
```suggestion
                    Extra... extra)
``?

Reply:
Actually,
```suggestion
                    ExtraArgs&&... extra)
```
so we can perfectly forward whatever was passed to the constructor.

Reply:
ok